compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -4779,6 +4779,152 @@ TR::Register *OMR::X86::TreeEvaluator::atomicorEvaluator(TR::Node *node, TR::Cod
+                                                                                                                                                                           
+   TR::LabelSymbol *startLabel = TR::LabelSymbol::create(cg->trHeapMemory(),cg);                                                                                           
+   startLabel->setStartInternalControlFlow();                                                                                                                              
+   TR::LabelSymbol *endLabel = TR::LabelSymbol::create(cg->trHeapMemory(),cg);                                                                                             
+   endLabel->setEndInternalControlFlow();                                                                                                                                  
+                    
+   TR::LabelSymbol *fallbackLabel = TR::LabelSymbol::create(cg->trHeapMemory(),cg);                                                                                    
+   TR::LabelSymbol *persistentFailureLabel = persistentFailureNode->getBranchDestination()->getNode()->getLabel();                                                         
+   TR::LabelSymbol *transientFailureLabel  = transientFailureNode->getBranchDestination()->getNode()->getLabel();     
+   TR::LabelSymbol *fallThroughLabel = fallThroughNode->getBranchDestination()->getNode()->getLabel();    
+
+   //if LabelSymbol isn't ready yet, generate a new one and assign to the node.
+   if(!fallThroughLabel){
+      fallThroughLabel = generateLabelSymbol(cg); 
+      fallThroughNode->getBranchDestination()->getNode()->setLabel(fallThroughLabel);
+   }         
+
+   if(!transientFailureLabel){
+       transientFailureLabel = generateLabelSymbol(cg); 
+       transientFailureNode->getBranchDestination()->getNode()->setLabel(transientFailureLabel);
+   }                                
+  
+   //in case user will make transientFailure goto persistenFailure, in which case the label will mess up at this point
+   //we'd better re-generate the label and set it to persistentFailure node again.
+   if(!persistentFailureLabel || persistentFailureLabel != persistentFailureNode->getBranchDestination()->getNode()->getLabel()){
+      persistentFailureLabel = generateLabelSymbol(cg);  
+      persistentFailureNode->getBranchDestination()->getNode()->setLabel(persistentFailureLabel);
+   }     
+       
+   TR::Register *accReg = cg->allocateRegister();                                                                                                                          
+   TR::RegisterDependencyConditions *endLabelConditions;
+   TR::RegisterDependencyConditions *fallThroughConditions = NULL;
+   TR::RegisterDependencyConditions *persistentConditions = NULL;
+   TR::RegisterDependencyConditions *transientConditions = NULL;
+
```

Comment:
This area of the code is not my specialty. @andrewcraik have you or a delegate reviewed this code already?